File: Part 4 - Clustering/Section 25 - Hierarchical Clustering/hc - ADONIS.py
# -*- coding: utf-8 -*-
"""
Created on Tue May 21 06:10:47 2019

@author: Adonis-Note
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

dataset = pd.read_csv('Mall_Customers.csv')
X = dataset.iloc[:, [3, 4]].values

# Using the dendrogram to find the optimal number of clusters
import scipy.cluster.hierarchy as sch
dendogram = sch.dendrogram(sch.linkage(X, method = 'ward'))   #linkage é o algoritmo para o hierarchical cluster.  Ward é o método que tenta minimizar a variância (distância) de cada cluster
plt.title('Dendrogram')
plt.xlabel('Customers')
plt.ylabel('Euclidean Distances')
plt.show()
#5 clusters

# Fitting the hierarchical clustering to the mall dataset
from sklearn.cluster import AgglomerativeClustering
hc = AgglomerativeClustering(n_clusters = 5, 
                             affinity = 'euclidean',
                             linkage = 'ward')
y_hc = hc.fit_predict(X)

# Visualising the clusters
plt.scatter(X[y_hc == 0, 0], X[y_hc == 0, 1], s = 100, c = 'red'    , label = 'Cluster 1')  # s define o tamanho do ponto no gráfico
plt.scatter(X[y_hc == 1, 0], X[y_hc == 1, 1], s = 100, c = 'blue'   , label = 'Cluster 2')
plt.scatter(X[y_hc == 2, 0], X[y_hc == 2, 1], s = 100, c = 'green'  , label = 'Cluster 3')
plt.scatter(X[y_hc == 3, 0], X[y_hc == 3, 1], s = 100, c = 'cyan'   , label = 'Cluster 4')
plt.scatter(X[y_hc == 4, 0], X[y_hc == 4, 1], s = 100, c = 'magenta', label = 'Cluster 5')
plt.title('Clusters of CLients')
plt.xlabel('Annual Income (k$)')
plt.ylabel('Spending Score (1-100)')
plt.legend()
plt.show()

File: Part 4 - Clustering/Section 25 - Hierarchical Clustering/hc - ADONIS.R
# Hierarchical Clustering

# Importing the mall dataset
dataset = read.csv('Mall_Customers.csv')
X = dataset[4:5]

# Using the dendrogram to find the optimal number of clusters
dendrogram = hclust(dist(X, method = 'euclidean'), method = 'ward.D')  # primeiro parâmetro "d" é a distance matrix of the dataset X, que é uma matriz que diz para cada par de customers, a euclidean distance entre os dois. Busca as coordenadas entre as variáveis x e y, e calcula a euclidean distance entre esses dois.
plot(dendrogram,
     main = paste('Dendrogram'),
     xlab = 'Customers',
     ylab = 'Euclidean distances')

# Fitting hierarchical clustering to the mall dataset
hc = hclust(dist(X, method = 'euclidean'), method = 'ward.D')  # primeiro parâmetro "d" é a distance matrix of the dataset X, que é uma matriz que diz para cada par de customers, a euclidean distance entre os dois. Busca as coordenadas entre as variáveis x e y, e calcula a euclidean distance entre esses dois.
y_hc = cutree(hc, 5)  # método utilizado para gerar o vector com a informação de qual cluster cada customer pertence

# Visualising the clusters
library(cluster)
clusplot(X,
         y_hc,
         lines = 0,
         shade = TRUE,
         color = TRUE,
         labels = 2,
         plotchar = FALSE,
         span = TRUE,
         main = paste('Clusters of Clients'),
         xlab = 'Annual Income',
         ylab = 'Spending Score')